src/omv/imlib/binary.c

```diff
@@ -260,54 +260,84 @@ static void imlib_b_and_line_op(image_t *img, int line, void *other, void *data,
+                        int p = dataPixel & otherPixel;
+                        IMAGE_PUT_BINARY_PIXEL_FAST(row, x, p);
                     }
                 }
             }
             break;
         }
         case PIXFORMAT_GRAYSCALE: {
-            uint8_t *data = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(img, line);
+            uint8_t *row = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(img, line);
 
             if (!mask) {
-                for (int i = 0, j = IMAGE_GRAYSCALE_LINE_LEN(img); i < j; i++) {
-                    data[i] &= ((uint8_t *) other)[i];
+                int x = 0;
+                #if (__ARM_ARCH > 6)
+                for (; (img->w - x) >= 4; x += 4) {
+                    int dataPixels = *((uint32_t *) (row + x));
+                    int otherPixels = *((uint32_t *) (((uint8_t *) other) + x));
+                    *((uint32_t *) (row + x)) = dataPixels & otherPixels;
+                }
+                #endif
+
+                for (; x < img->w; x++) {
+                    int dataPixel = IMAGE_GET_GRAYSCALE_PIXEL_FAST(row, x);
+                    int otherPixel = IMAGE_GET_GRAYSCALE_PIXEL_FAST(((uint8_t *) other), x);
+                    int p = dataPixel & otherPixel;
+                    IMAGE_PUT_GRAYSCALE_PIXEL_FAST(row, x, p);
                 }
             } else {
-                for (int i = 0, j = img->w; i < j; i++) {
-                    if (image_get_mask_pixel(mask, i, line)) {
-                        IMAGE_PUT_GRAYSCALE_PIXEL_FAST(data, i,
-                                                       (IMAGE_GET_GRAYSCALE_PIXEL_FAST(data, i)
-                                                        & IMAGE_GET_GRAYSCALE_PIXEL_FAST(((uint8_t *) other), i)));
+                for (int x = 0; x < img->w; x++) {
+                    if (image_get_mask_pixel(mask, x, line)) {
+                        int dataPixel = IMAGE_GET_GRAYSCALE_PIXEL_FAST(row, x);
+                        int otherPixel = IMAGE_GET_GRAYSCALE_PIXEL_FAST(((uint8_t *) other), x);
+                        int p = dataPixel & otherPixel;
+                        IMAGE_PUT_GRAYSCALE_PIXEL_FAST(row, x, p);
                     }
                 }
             }
             break;
         }
         case PIXFORMAT_RGB565: {
-            uint16_t *data = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(img, line);
+            uint16_t *row = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(img, line);
 
             if (!mask) {
-                for (int i = 0, j = IMAGE_RGB565_LINE_LEN(img); i < j; i++) {
-                    data[i] &= ((uint16_t *) other)[i];
+                int x = 0;
+                #if (__ARM_ARCH > 6)
+                for (; (img->w - x) >= 2; x += 2) {
+                    int dataPixels = *((uint32_t *) (row + x));
+                    int otherPixels = *((uint32_t *) (((uint16_t *) other) + x));
+                    *((uint32_t *) (row + x)) = dataPixels & otherPixels;
+                }
```

Comment:
I think this is a bit convoluted for no good reason. All 4 bytes loops should be the same, i.e:
```c
                for (; (img->w - x) >= 4; x += 4) {
                    uint32_t p0 = *((uint32_t *) (row0 + x));
                    uint32_t p1 = *((uint32_t *) (row1 + x));
                    *((uint32_t *) (row + x)) = p0 & p1;
                }
```
In fact, for some reason this is slightly faster.

Reply:
The loops working on RGB565 images have to be slightly different than the ones on GRAYSCALE images as the pointers are different.

Reply:
Those two should be same, how is it different or how can it fail ?

Reply:
Ah yes I see, this should be `for (; (img->w - x) >= 2; x += 2) {`